proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+exceptions.
+
+One issue is that both sides need to know if an exception was thrown by the
+other, because cleanup may need to be performed.
+
+Another problem is that WebAssembly doesn't have direct access to the host VM's
+memory.  As a result, WebAssembly defers the handling of exceptions to the host
+VM.
+
+To access exceptions, WebAssembly provides instructions to check if the
+exception is one that WebAssembly understands. If so, the data of the
+WebAssembly exceptions's data is extracted and copied onto the stack, allowing
+succeeding instructions to process the data.
+
+Lastly, exception lifetimes must be maintained by the host VM, so that it can
+collect and reuse the memory used by exceptions. This implies that the host must
+know where exceptions are stored, so that it can determine when an exception can
+be garbage collected.
+
+This also implies that the host VM must provide a garbage collector for
+exceptions.  For host VMs that have garbage collection (such as JavaScript),
+this is not a problem.
+
+However, not all host VMs may have a garbage collector. For this reason,
+WebAssembly exceptions are designed to allow the use of reference counters to
+perform the the garbage collection in the host VM.
+
+To do this, WebAssembly exceptions are immutable once created, to avoid cyclic
+data structures that can't be garbage collected. It also means that exceptions
+can't be stored into linear memory. The rationale for this is twofold:
+
+* For security. Loads
+  and stores do not guarantee that the data read was of the same type as
+  stored. This allows spoofing of exception references that may allow a
+  WebAssembly module to access data it should not know in the host VM.
+  
+* The host VM does not know the layout of data in linear memory, so it can't
+  find places where exception references are stored.
+
+Hence, while an exception reference is a new first class type, this proposal
+disallows their usage in linear memory.
+
+A WebAssembly exception is created by the `except` instruction. Once an
+exception is created, you can throw it with the `throw` instruction. Thrown
+exceptions are handled as follows:
+
+1. They can be caught by a catch block in an enclosing try block of a function
+   body. The caught exception is pushed onto the stack.
+
+1. Throws not caught within a function body continue up the call stack, popping
+   call frames, until an enclosing try block is found.
+   
+1. If the call stack is exhausted without any enclosing try blocks, the host VM
+   defines how to handle the uncaught exception.
+
+### Exceptions
+
+An `exception` is an internal construct in WebAssembly that is maintained by the
+host. WebAssembly exceptions (as opposed to host exceptions) are defined by a
+new `exception section` of a WebAssembly module. The exception section is a list
+of exception types, from which exceptions can be created.
+
+Each exception type has a `type signature`. The type signature defines the list
```

Comment:
Playing advocate of the devil here: since in the general case a C++ exception type cannot be represented as a self-contained wasm signature (e.g. if it contains a string) I'd expect the front-end to often or always opt for a signature that is just a single linear memory index. Similarly, a host exception will likely often be a single anyref, once we get that. This in turn does not explain to me why we need the generality of a type signature, other than, "why not, may be nice to have".

Reply:
Its more of the latter. This topic was brought up in many earlier CG meetings and the general consensus was that exceptions should not just be C++ specific, but more general so that other languages could be translated down to WebAssembly.